### PR TITLE
[Proposal] App Catalog Cleanup Reusable Workflow

### PR DESCRIPTION
**Pull request by ljakimczuk**
### TL;DR

This PR introduces a custom GHA workflow that stands the interface to run the ACCT against the catalog repositories that references it. It follows the [Reusable Workflows](https://docs.github.com/en/actions/learn-github-actions/reusing-workflows) guide, which although being in beta, seems to work quite fine. A small issue described below has been spotted, but it also seems quite easy to overcome.

### Spotted issues

Passing the GH token between the caller and called workflows is the only problem spotted so far. The reusable workflow may define `secrets` the other workflows can pass by upon calling it, but this mechanism does not seem to work with the GH tokens. Though, they can still be accessed by the reusable workflow using the `github` context, see `Create Pull Request` step.

### Why reusable workflow?

It seems the best fit the use case. Other natural choice would be a dedicated GH Action, but this repository feels to be rather dedicated to the ACCT application and its releases, so turning it into the GHA repository doesn't feel right, neither creating an extra repository for that purpose. Workflow is in addition more flexible, action corresponds to a single step, while we can craft any number of them in a workflow.

### Referencing workflow

Catalog repos can reference this workflow in their own workflows, see the [Control Plane Test Catalog example](https://raw.githubusercontent.com/giantswarm/control-plane-test-catalog/feature/cleanup-workflow/.github/workflows/cleanup.yaml), or the snippet below.

```yaml
name: Control Plane Test Catalog Cleanup

on:
  schedule:
    - cron: "0 0 1 * *"

jobs:
  cleanup:
    uses: giantswarm/app-catalog-cleanup-tool/.github/workflows/cleanup.yaml@feature/cleanup-workflow
    with:
      delete-before: "1 month"
      dry-run: false
```

### Commiting changes

Creating a PR seems like an obvious way of commiting cleanup. We could easily revert it in case of any issues or cleaning up too much. In result, the workflow, after running the ACCT opens a PR with changes.

### Yet to be done

**Merging**: how would we like to merge the cleanup PR. It may depend on the catalog type, like for example we could use auto-merge on `-test` catalogs and approvals on the production catalogs.

**PR Notifications**: for the repositories without auto-merge, what would be the notification channel to inform owners of the catalog repository about the PR ready for merge? Pinging on the Slack's channel maybe?

**General notification**: do we want somehow track the entries we remove, except the GHA stdout and PR diff itself? Like for example in form of metrics or logs stored somewhere?

### CODE REVIEW

File: .github/workflows/cleanup.yaml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            echo '::set-output name=condition::--delete-before "${{ inputs.delete-before }}"'
+          elif [[ ! -z "${{ inputs.limit-number }}" ]]
+          then
+            echo '::set-output name=condition::--limit-number ${{ inputs.limit-number }}'
+          else
+            echo '::set-output name=condition::'
+          fi
+
+      - name: Execute ACC tool
+        uses: docker://quay.io/giantswarm/app-catalog-cleanup-tool:latest
```
Comment (rossf7) [inline .github/workflows/cleanup.yaml:82]:
```suggestion
        uses: docker://quay.io/giantswarm/app-catalog-cleanup-tool:0.1.2
```

Same here. Let's use the latest release but we will need to find a way to update this.

File: .github/workflows/cleanup.yaml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      dry-run:
+        description: "Enable dry-run mode"
+        default: true
+        required: false
+        type: boolean
+
+jobs:
+  cleanup:
+    name: Run the ACC tool
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (ljakimczuk) [reply .github/workflows/cleanup.yaml]: makes sense, have changed that

**Comment by ljakimczuk** [comment]
@rossf7 small update from my, `renovatebot` does seem to work fine with the reusable workflows after taking closer look at the code, so it should be fine.

File: .github/workflows/cleanup.yaml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+      - name: Create Pull Request
+        if: ${{ ! inputs.dry-run }}
+        uses: peter-evans/create-pull-request@v3
+        with:
+          token: ${{ github.token }}
+          title: Periodic App Catalog Cleanup
+          commit-message: Periodic App Catalog Cleanup
+          branch: cleanup/acct-run-${{ github.run_id }}
+          reviewers: giantswarm/team-honeybadger
```
Comment (rossf7) [inline .github/workflows/cleanup.yaml:94]: There is also `team-reviewers` does it matter which one we use?

https://github.com/marketplace/actions/create-pull-request

Comment (ljakimczuk) [reply .github/workflows/cleanup.yaml:94]: good point, probably the `team-reviewers` is the one we need, will change that
